sensor/common/sensor/central_receiver_test.go

```diff
@@ -0,0 +1,418 @@
+	s.finished = &sync.WaitGroup{}
+	s.finished.Add(1)
+}
+
+func (s *centralReceiverSuite) TearDownTest() {
+	if s.receiver != nil {
+		s.receiver.Stop()
+		s.finished.Wait()
+	}
+	goleak.AssertNoGoroutineLeaks(s.T())
+}
+
+func Test_CentralReceiverSuite(t *testing.T) {
+	suite.Run(t, new(centralReceiverSuite))
+}
+
+func (s *centralReceiverSuite) Test_StreamContextCancelShouldStopFlow() {
+	s.receiver = NewCentralReceiver(s.finished)
+	streamContext, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	s.mockClient.EXPECT().Context().AnyTimes().Return(streamContext).AnyTimes()
+
+	s.mockClient.EXPECT().Recv().Times(1).DoAndReturn(func() (*central.MsgToSensor, error) {
+		cancel()
+		s.T().Logf("Canceled context")
+		return testMsg, nil
+	})
+
+	s.receiver.Start(s.mockClient)
+	s.finished.Wait()
+	s.EqualError(s.receiver.Stopped().Err(), "context canceled")
+}
+
+func (s *centralReceiverSuite) Test_RecvErrorShouldStopFlow() {
+	s.receiver = NewCentralReceiver(s.finished)
+	s.mockClient.EXPECT().Context().AnyTimes().Return(context.Background()).AnyTimes()
+
+	s.mockClient.EXPECT().Recv().Times(1).DoAndReturn(func() (*central.MsgToSensor, error) {
+		return nil, errors.New("some error")
+	})
+
+	s.receiver.Start(s.mockClient)
+	s.finished.Wait()
+	s.EqualError(s.receiver.Stopped().Err(), "some error")
+}
+
+func (s *centralReceiverSuite) Test_EofShouldStopFlowWithNoError() {
+	s.receiver = NewCentralReceiver(s.finished)
+	s.mockClient.EXPECT().Context().AnyTimes().Return(context.Background()).AnyTimes()
+
+	s.mockClient.EXPECT().Recv().Times(1).DoAndReturn(func() (*central.MsgToSensor, error) {
+		return nil, io.EOF
+	})
+
+	s.receiver.Start(s.mockClient)
+	s.finished.Wait()
+	s.NoError(s.receiver.Stopped().Err())
+}
+
+func (s *centralReceiverSuite) Test_SlowComponentDoesNotBlockOthers() {
+	fastTick := make(chan struct{})
+	close(fastTick)
+	fastComponent := &testSensorComponent{
+		t:          s.T(),
+		name:       "fast",
+		tick:       fastTick,
+		responsesC: make(chan *message.ExpiringMessage),
+	}
+	blockedChan := make(chan struct{})
+	blockingComponent := &testSensorComponent{
+		t:          s.T(),
+		name:       "blocked",
+		tick:       blockedChan,
+		responsesC: make(chan *message.ExpiringMessage),
+	}
+
+	components := []common.SensorComponent{fastComponent, blockingComponent}
+	s.receiver = NewCentralReceiver(s.finished, components...)
+
+	s.mockClient.EXPECT().Context().AnyTimes().Return(context.Background()).AnyTimes()
+
+	msgCount := 3
+	s.mockClient.EXPECT().Recv().MinTimes(msgCount).DoAndReturn(func() (*central.MsgToSensor, error) {
+		return testMsg, nil
+	})
+
+	s.receiver.Start(s.mockClient)
+
+	for i := 0; i < msgCount; i++ {
+		select {
+		case <-fastComponent.ResponsesC():
+			s.T().Logf("Fast component processed the message")
+		case <-blockingComponent.ResponsesC():
+			assert.FailNow(s.T(), "blocked component should not process the message")
+		}
+	}
+
+	s.receiver.Stop()
+	s.finished.Wait()
```

Comment:
should we stop the components too? Also `blockedChan` is never closed, should we close it?